### PR TITLE
Rearchitect test generator

### PR DESCRIPTION
**Pull request by marionebl**
* integrate `ocp-indent` into `test_gen` directly
* migrate from custom templating to `ocaml-mustache`
* allow for arbitrary file templates per exercise
* Add special_case transformations on `case` level (as opposed to `input`)
* move `tool/test-generator` to `test_generator`

### CODE REVIEW

File: test-generator/lib_generator/canonical_data.ml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  )    
+  in
+  {
+    version;
+    exercise;
+    comments = (try (mem "comments") |> Util.to_list |> List.map ~f:Util.to_string with _ -> []);
+    cases = (mem "cases") |> Util.to_list |> sanitize_cases
+  }
+
+let rec yo_to_ez (j: Yojson.Basic.t): Ezjsonm.value =
```
Comment (sshine) [inline test-generator/lib_generator/canonical_data.ml]: I'm unsure what role the Ezjson library has when there's also Yojson, is it a dependency thing in the template library?

**Comment by sshine** [comment]
Does this PR leave anything missing in #284?

File: exercises/beer-song/test.ml [hunk 2]
```diff
@@ -54,8 +55,10 @@ let lyrics_tests = [
...
 
+
+
 let () =
   run_test_tt_main (
     "beer song tests" >:::
-    List.concat [verse_tests; lyrics_tests]
-  )
\ No newline at end of file
+    List.concat [verse; lyrics; ]
```
[the remark this answers is not in the document]
Comment (marionebl) [reply exercises/beer-song/test.ml]: Fixed via d9054c724a2127548ca98680c2f4542d0a057117

File: exercises/forth/test.ml [hunk 3]
```diff
@@ -154,13 +156,13 @@ let () =
...
       dup_tests; 
-      drop_tests;
+      drop_tests; 
       swap_tests; 
       over_tests; 
-      user_defined_words_tests;
-      case_insensitivity;
+      user_defined_words_tests; 
+      case_insensitivity_tests; 
     ]
```
[the remark this answers is not in the document]
Comment (marionebl) [reply exercises/forth/test.ml]: Fixed via d9054c724a2127548ca98680c2f4542d0a057117

File: exercises/run-length-encoding/test.ml [hunk 4]
```diff
@@ -6,43 +7,44 @@ let ae exp got _test_ctxt = assert_equal exp got ~printer:Fn.id
...
   "lower case string" >::
-  ae "aabbbcccc" (decode "2a3b4c");
+  ae "aabbbcccc" ("2a3b4c" |> decode );
 ]
-
-
 let encode_and_then_decode_tests = [
   "encode followed by decode gives original string" >::
-  ae "zzz ZZ  zZ" (encode "zzz ZZ  zZ" |> decode);
+  ae "zzz ZZ  zZ" ("zzz ZZ  zZ" |> encode |> decode );
```
[the remark this answers is not in the document]
Comment (marionebl) [reply exercises/run-length-encoding/test.ml]: Fixed via 6679c7797369f807759ba94d7960f6ad31fdac07

File: templates/beer-song/test.ml.tpl [hunk 5]
```diff
@@ -23,26 +24,22 @@ let zip s1 s2 =
...
-]
-(* END SUITE *)
+
+{{/cases}}
 
 let () =
   run_test_tt_main (
     "beer song tests" >:::
-      List.concat [verse_tests; lyrics_tests]
+      List.concat [{{#cases}}{{{description}}}; {{/cases}}]
```
[the remark this answers is not in the document]
Comment (marionebl) [reply templates/beer-song/test.ml.tpl]: Fixed via d9054c7

File: templates/forth/test.ml.tpl [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+
+{{/cases}}
+
+let () =
+  run_test_tt_main (
+    "forth tests" >:::
+      List.concat [
+        {{#cases}}
+          {{slug}}_tests;
```
[the remark this answers is not in the document]
Comment (marionebl) [reply templates/forth/test.ml.tpl]: Fixed via 2bd62ab and 5da13cd

File: templates/run-length-encoding/test.ml.tpl [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+]
+{{/cases}}
+
+
+let () =
+  run_test_tt_main (
+    "run length encoding tests" >:::
+      List.concat [
+        {{#cases}}
+          {{slug}}_tests;
```
[the remark this answers is not in the document]
Comment (marionebl) [reply templates/run-length-encoding/test.ml.tpl]: Fixed

File: test-generator/lib_generator/canonical_data.ml (hunk 1 above)
Comment (marionebl) [reply test-generator/lib_generator/canonical_data.ml]: Yes, `ocaml-mustache` uses `Ezjson.t` as data type, thus this conversion and the introduction of the library into the dependency list

**Comment by marionebl** [comment]
> Does this PR leave anything missing in #284?

This adds versioning to all exercises that generate tests from `problem-specifications`. Exceptions:

- custom-set
- grade-school
- hangman
- hexadecimal
- list-ops
- meetup
- nucleotide-count
- react
- rna-transcription
- robot-name
- zipper

Also the new test_gen does not create `.meta/version` files yet

**Comment by sshine** [comment]
Since this PR doesn't add `.meta/version` to the generator, I wonder if we should leave this as a "good first issue" for the upcoming hacktoberfest. I think this would be very approachable considering how much you've modernized the test generator already.

**Comment by sshine** [comment]
I've updated #284 to reflect the changes this PR makes and what's still missing wrt. versioning.

**Comment by sshine** [comment]
I've removed this point and re-inserted it here: 

> For the sake of merging soon, the only thing I'd like to have addressed is that I am not sure why `.editorconfig` was added to this PR. What does it have to do with the test generator?

**Comment by marionebl** [comment]
> I've removed this point and re-inserted it here:

>> For the sake of merging soon, the only thing I'd like to have addressed is that I am not sure why .editorconfig was added to this PR. What does it have to do with the test generator?


Ah I didn't catch this. I added `.editorconfig` as a semi-automated way to address the multitude of trailing whitespace issues that popped up in the review. Also I had the lofty idea to write a `editorconfig` implementation in OCaml and integrate it into `test_gen`. After looking into the complexity of existing implementations I backed of from that part of the plan (for now).

**Comment by marionebl** [comment]
> I see that your current solution wrt. having both generated oUnit and manually written QCheck is to add QCheck exercises to the test.ml.tpl template file [...]

> I think this is very acceptable; [...] I wonder if we should consider [...] whether to document this part of the test generator.

Yes, let's document the updated mechanics of `test_gen`. I suspect we'll find some opportunities to make the project layout a bit easier to understand while documenting - I ran into mental overhead mainly caused by having to relate the source files in `templates`, `exercises` and `problem-specifications/exercises` while working on a single exercise.

I repeatedly thought it might be easier to have all sources reside in `templates` and use `exercises` as "artifact storage" exclusively.  

> I don't recall if you have made any CI checks that run the test generator and sees if there is a discrepancy? Providing a meaningful CI error would be very useful if contributors in the future think they can address the generated test suites directly, and it also asserts that the test generator can be built.

Yes, those checks are in place: 

**1. Consistency check:**
https://github.com/exercism/ocaml/blob/3d241631079ed6814065527dd03105df555e6d31/.travis/.travis-ci.sh#L18-L24

**2. `test_gen` run
https://github.com/exercism/ocaml/blob/3d241631079ed6814065527dd03105df555e6d31/.travis/.travis-ci.sh#L13

**Comment by sshine** [comment]
Hooray!

**Pull request merged**